gcloud/search/document.py

```diff
@@ -0,0 +1,303 @@
+
+    See:
+    https://cloud.google.com/search/reference/rest/v1/projects/indexes/documents
+
+    :type name: string
+    :param name: the name of the document
+
+    :type index: :class:`gcloud.search.index.Index`
+    :param index: the index to which the document belongs.
+
+    :type rank: int
+    :param rank: the default rank for ordering the document.
+    """
+    def __init__(self, name, index, rank=None):
+        self.name = name
+        self.index = index
+        self.rank = rank
+        self.fields = {}
+
+    @classmethod
+    def from_api_repr(cls, resource, index):
+        """Factory:  construct a document given its API representation
+
+        :type resource: dict
+        :param resource: document resource representation returned from the API
+
+        :type index: :class:`gcloud.search.index.Index`
+        :param index: Index holding the document.
+
+        :rtype: :class:`gcloud.search.document.Document`
+        :returns: Document parsed from ``resource``.
+        """
+        name = resource['docId']
+        rank = resource.get('rank')
+        instance = cls(name, index, rank)
+        instance._parse_fields_resource(resource)
+        return instance
+
+    def _parse_value_resource(self, resource):
+        """Helper for _parse_fields_resource"""
+        if 'stringValue' in resource:
+            string_format = resource.get('stringFormat')
+            language = resource.get('lang')
+            value = resource['stringValue']
+            return StringValue(value, string_format, language)
+        if 'numberValue' in resource:
+            value = resource['numberValue']
+            if isinstance(value, six.string_types):
+                if '.' in value:
+                    value = float(value)
+                else:
+                    value = int(value)
+            return NumberValue(value)
+        if 'timestampValue' in resource:
+            stamp = resource['timestampValue']
+            value = datetime.datetime.strptime(stamp, _RFC3339_MICROS)
+            value = value.replace(tzinfo=UTC)
+            return TimestampValue(value)
+        if 'geoValue' in resource:
+            lat_long = resource['geoValue']
+            lat, long = [float(coord.strip()) for coord in lat_long.split(',')]
+            return GeoValue((lat, long))
+        raise ValueError("Unknown value type")
+
+    def _parse_fields_resource(self, resource):
+        """Helper for from_api_repr, create, reload"""
+        self.fields.clear()
+        for field_name, val_obj in resource.get('fields', {}).items():
+            field = self.field(field_name)
+            for value in val_obj['values']:
+                field.values.append(self._parse_value_resource(value))
+
+    @property
+    def path(self):
+        """URL path for the document's APIs"""
+        return '%s/documents/%s' % (self.index.path, self.name)
+
+    def field(self, name):
+        """Construct a Field instance.
```

Comment:
It is a factory, yes.

The document holds onto the fields, and sends them with its resource repr in `create`.